### PR TITLE
Async building

### PR DESCRIPTION
**Pull request by geoffp**
# Overview

This is a PR to track the first batch of asynchronous pattern building reform!

The purpose of this work is twofold:
* **Be able to use asynchronous APIs** within engines, plugins, and main code (I'm looking at you, Webpack)
* **Be faster at building**. I haven't done any formal benchmarking, but very early results seem to show a little speed benefit. I'm hoping this will increase as we start to knock down synchronous filesystem calls, in particular. In future:
  * Increase the use of Node async APIs, especially for IO-bound operations
  * Enable parallelism with surrounding build tools like gulp that are happy to get super async if you let them.

The async strategy is focused on Promises. We've made the choice to support Node 4.4 LTS for now, but this strategy does let us move easily to async/await (since that syntax is just sugar around Promises) whenever the time is right.

# Interesting things that happened
The whole `processAllPatternsIterative()` build step was impossible to make async until I split it into loading and processing phases. **Loading** them all first in a separate loop from **processing** takes the race conditions away. I believe this has the side effect that you can now include an organism in an atom if you really want to, but please don't.

# Testing
You'll need the **async-building** branch of **edition-node-gulp** to use this. I've been testing it with starterkit-mustache-demo.

# Impact
The whole control flow of PL needs to be async to enable this -- but I don't think that's a bad thing at all -- it's just work, and probably lots of fixes to unit tests. @raphaelokon should be aware of this -- eventually the CLI will need to get all .then() because of this.

### CODE REVIEW

**Comment by raphaelokon** [comment]
Hi @geoffp :-) Just noticed you mentioned me here. The first cut of CLI is there … I am looking forward to help on core as well. The cli uses coroutines to yield out Promises. So for me it is just prepending the call to core with a `yield`. But asynchrony is key to make the core fly.

**Comment by geoffp** [comment]
@raphaelokon, I admit, I've been a little dismissive of coroutines, so I don't know much about them. I'm interested in what you use them for and how they compare to promises.

Also, @bmuenzenmeyer, I see I have lots of unit tests to clean up in this branch. For now, this PR is up early, mostly for code review, but if we're all aligned on 3.0, I suppose we can merge some of this stuff whenever the tests start passing.

**Comment by geoffp** [comment]
Okay, now I need to figure out why I'm passing all my local unit tests but Travis is still being grouchy.

**Comment by geoffp** [comment]
@bmuenzenmeyer -- this is now passing tests. Would you review before this gets much bigger? Next round is the loading process.

**Comment by bmuenzenmeyer** [comment]
@geoffp will do. should have some thoughts by Monday

File: package.json [hunk 1]
```diff
@@ -4,8 +4,11 @@
   "version": "2.6.2",
   "main": "./core/lib/patternlab.js",
   "dependencies": {
+    "async": "^2.1.2",
+    "dive": "^0.5.0",
     "diveSync": "^0.3.0",
```
Comment (bmuenzenmeyer) [inline package.json]: #389  exists to attempt to merge a couple dependencies. do you see us still being able to do so with the addition of the true async dive?

File: core/lib/pattern_assembler.js [hunk 2]
```diff
@@ -45,6 +45,7 @@ var pattern_assembler = function () {
         return patternlab.patterns[i];
       }
     }
+    console.trace("Here I am!");
```
Comment (bmuenzenmeyer) [inline core/lib/pattern_assembler.js]: curious as to why i dont see this message show up in my output. remove when you no longer need this

File: core/lib/patternlab.js [hunk 3]
```diff
@@ -35,18 +36,43 @@ function buildPatternData(dataFilesPath, fsDep) {
...
-      //log any errors
-      if (err) {
-        console.log(err);
-        return;
-      }
-      pattern_assembler.process_pattern_iterative(path.relative(patterns_dir, file), patternlab);
-    }
-  );
+function processAllPatternsIterative(pattern_assembler, patterns_dir, patternlab){
+  const async = require('async');
```
Comment (bmuenzenmeyer) [inline core/lib/patternlab.js]: is this dead code?

File: core/lib/patternlab.js [hunk 4]
```diff
@@ -35,18 +36,43 @@ function buildPatternData(dataFilesPath, fsDep) {
...
+        // can load them asynchronously and be sure we know about all
+        // of them before we start lineage hunting, for
+        // example. Incidentally, this should also allow people to do
+        // horrifying things like include a page in a atom. But
+        // please, if you're reading this: don't.
+
+        // NOTE: sync for now
+        pattern_assembler.load_pattern_iterative(path.relative(patterns_dir, file), patternlab);
+      },
+      resolve
```
Comment (bmuenzenmeyer) [inline core/lib/patternlab.js]: I was trying to wrap my head around what this means. My understanding is this is the rejection function - which you say to call the `resolve` function, yes?

File: core/lib/pseudopattern_hunter.js [hunk 5]
```diff
@@ -1,81 +1,110 @@
...
-    var needle = currentPattern.subdir + '/' + currentPattern.fileName + '~*.json';
-    var pseudoPatterns = glob.sync(needle, {
-      cwd: paths.source.patterns,
-      debug: false,
-      nodir: true
-    });
-
-    if (pseudoPatterns.length > 0) {
-      for (var i = 0; i < pseudoPatterns.length; i++) {
+const glob = require('glob'),
```
Comment (bmuenzenmeyer) [inline core/lib/pseudopattern_hunter.js]: thanks for moving stuff like this out of the function - i have slowly been doing so too!

File: core/lib/pseudopattern_hunter.js [hunk 6]
```diff
@@ -1,81 +1,110 @@
...
+      { cwd: paths.source.patterns, debug: false, nodir: true },
+      function (err, matches) {
+        if (err) { reject(err); }
+        resolve(matches);
+      }
+    );
+  });
+}
+
+function createPseudoPatternObject(variantFileData, pattern, pseudoPattern, patternlab) {
```
Comment (bmuenzenmeyer) [inline core/lib/pseudopattern_hunter.js]: I really like what you did with this file to make each function have a clearer separation of concerns, but not to a zealous extent

File: core/lib/pseudopattern_hunter.js [hunk 7]
```diff
@@ -1,81 +1,110 @@
...
+              currentPattern,
+              pseudoPatternPath,
+              patternlab
+            );
+
+            //process the companion markdown file if it exists
+            processPseudoPattern(patternVariant, patternlab);
+          })
+          .catch(plutils.reportError(
+            'There was an error processing the pseudopattern' + pseudoPatternPath
```
Comment (bmuenzenmeyer) [inline core/lib/pseudopattern_hunter.js]: space needed after `pseudopattern`

Comment (raphaelokon) [inline core/lib/pseudopattern_hunter.js]: Use ES6 template strings ^^ instead of concat operator

File: core/lib/patternlab.js (hunk 4 above)
Comment (raphaelokon) [inline core/lib/patternlab.js]: Nope, this is the complete handler according to `dive` API: `dive(directory[, options], action[, complete]);`

Comment (bmuenzenmeyer) [inline core/lib/patternlab.js]: :+1:

File: core/lib/pseudopattern_hunter.js (hunk 6 above)
Comment (geoffp) [inline core/lib/pseudopattern_hunter.js]: Thank you! I try to make it obvious just by reading the code of a function like this that its basic procedure is correct, even if the implementations of those sub-functions might not be.

File: core/lib/patternlab.js (hunk 3 above)
Comment (geoffp) [inline core/lib/patternlab.js]: Yup. I need to take that out. I'm not currently using async for anything. I thought I would need it, but it turns out that I don't have a need for it so far. And in any case, it's not very Promise-y.

**Comment by geoffp** [comment]
So, @bmuenzenmeyer -- if you're cool with where this is going, I've got another batch of work ready for review. What would you like to do for a release / branching strategy?

Off the top of my head: we could use this branch as the "master" for this line of work, or we could merge this to dev (comitting to it, I suppose), and continue merging batches of work to dev that way.

File: core/lib/pseudopattern_hunter.js (hunk 7 above)
Comment (geoffp) [inline core/lib/pseudopattern_hunter.js]: All string concats in this module have now been migrated.

File: core/lib/pattern_assembler.js (hunk 2 above)
Comment (geoffp) [inline core/lib/pattern_assembler.js]: Will check this out.

**Comment by bmuenzenmeyer** [comment]
> if you're cool with where this is going, I've got another batch of work ready for review.

Yes!

> What would you like to do for a release / branching strategy?

I'd like to plan out what we do similar to the CLI work: https://github.com/pattern-lab/patternlab-node/projects/3 (which mostly is major revs to editions whereas this is breaking change to core)

I have created an async project here to start the planning: https://github.com/pattern-lab/patternlab-node/projects/2

Let me know if this format does not feel right. We can certainly try anything else too.

> we could use this branch as the "master" for this line of work

I'd be fine with this - effectively the same as our `dev-2.0` branch we had for awhile. I like to keep `dev` in a near-releasable state so I think keeping this out of there would be smart.

**Comment by geoffp** [comment]
That sounds right to me. I'll have another PR ready today.

**Comment by geoffp** [comment]
Ack, the latest round of changes to dev are playing havoc with this. It seems to want to blow away large chunks of patternlab.js for no reason. Working on it now.

**Comment by geoffp** [comment]
Think I fixed it. Now I'll see if I can update `async-building-pattern-loading`.

**Comment by bmuenzenmeyer** [comment]
> Ack, the latest round of changes to dev are playing havoc with this. It seems to want to blow away large chunks of patternlab.js for no reason. Working on it now.

hopefully it wasn't too bad! I toyed with the idea of doing my work in this branch - but ultimately thought it was unrelated (perhaps not!)

**Comment by geoffp** [comment]
No, it was mostly just git being (surprisingly) dumb about a my refactoring of the main build function and @raphaelokon's refactoring of utilities.js. Raphael's was easy to fix, mine was harder. But it's done!

**Comment by raphaelokon** [comment]
@geoffp Great to hear that 😛

**Comment by geoffp** [comment]
@bmuenzenmeyer, @tburny, @raphaelokon, I've just done part of the merge. I believe this branch is passing tests now. In my sloppy benchmarks during development, it seems to be roughly as fast as it was before on the small tree I'm testing with.

I'd appreciate any testing; I don't have much confidence yet that the UI build order is correct due to some vague but lingering memories from two months ago.

The bleeding edge of this work is currently on `async-rendering`. That one is still broken.

**Comment by bmuenzenmeyer** [comment]
I will try to test this tomorrow during the day. Hitting my cognitive wall for tonight.

**Comment by bmuenzenmeyer** [comment]
@geoffp this seems good when run as part of `dev-3.0` - even pattern `order` is adhered to.

considering this is now on dev-3.0, should this be closed, or are their specifics you still want reviewed?
I see two failing tests when running `npm test`

```

  483 passing (23s)
  2 pending
  2 failing

  1) test/engine_handlebars_tests.js @partial-block template should render without throwing (@geoffp repo issue #3) Error: The partial @partial-block could not be found:
     Error: The partial @partial-block could not be found
      at Object.Pattern.render (core/lib/object_factory.js:101:26)
      at test/engine_handlebars_tests.js:262:38

  2) test/lineage_hunter_tests.js cascade_pattern_states sets the pattern state on any lineage patterns reverse lineage should be equal:

      Error: should be equal
      + expected - actual

      -complete
      +inreview

      at test/lineage_hunter_tests.js:286:8
      at Object.<anonymous> (test/lineage_hunter_tests.js:262:5)
```

I will try to look at these tonight

**Comment by bmuenzenmeyer** [comment]
@geoffp 

poke regarding my previous comment

i've fixed all but one unit test, which I will look into now.

**Comment by geoffp** [comment]
@bmuenzenmeyer yes, I think this can be closed now. I have another async branch that contains some appealing patterns, but became too risky and tried to do too much at once, so I'd like to proceed by bringing over ideas from that branch one at a time, with a short-term focus on async rendering, since that's what's blocking Webpack adoption in the React engine.

**Comment by bmuenzenmeyer** [comment]
sounds great @geoffp please let me know how i can help further!

**Pull request closed without merging**